Doc/c-api/typeobj.rst

```diff
@@ -1497,20 +1554,101 @@ and :c:data:`PyType_Type` effectively act as defaults.)
+
+      Referents (direct and indirect) might have already been cleared; they are
+      not guaranteed to be in a consistent state.
+
+   .. warning::
+
+      The :c:member:`~PyTypeObject.tp_clear` function can be called from any
+      thread, although the :term:`GIL` will be held.
+
+   .. warning::
+
+      An object is not guaranteed to be automatically cleared before its
+      destructor (:c:member:`~PyTypeObject.tp_dealloc`) is called.
+
+   This function differs from the destructor
+   (:c:member:`~PyTypeObject.tp_dealloc`) in the following ways:
+
+   * The purpose of clearing an object is to remove references to other objects
+     that might participate in a reference cycle.  The purpose of the
+     destructor, on the other hand, is a superset: it must release *all*
+     resources it owns, including references to objects that cannot participate
+     in a reference cycle (e.g., integers) as well as the object's own memory
+     (by calling :c:member:`~PyTypeObject.tp_free`).
+   * When :c:member:`!tp_clear` is called, other objects might still hold
+     references to the object being cleared.  Because of this,
+     :c:member:`!tp_clear` must not deallocate the object's own memory
+     (:c:member:`~PyTypeObject.tp_free`).  The destructor, on the other hand,
+     is only called when no (strong) references exist, and as such, must
+     safely destroy the object itself by deallocating it.
+   * :c:member:`!tp_clear` might never be automatically called.  An object's
+     destructor, on the other hand, will be automatically called some time
+     after the object becomes unreachable (i.e., either there are no references
+     to the object or the object is a member of a :term:`cyclic isolate`).
+
+   No guarantees are made about when, if, or how often Python automatically
+   clears an object, except:
+
+   * Python will not automatically clear an object if it is reachable, i.e.,
+     there is a reference to it and it is not a member of a :term:`cyclic
+     isolate`.
+   * Python will not automatically clear an object if it has not been
+     automatically finalized (see :c:member:`~PyTypeObject.tp_finalize`).  (If
+     the finalizer resurrected the object, the object may or may not be
+     automatically finalized again before it is cleared.)
+   * If an object is a member of a :term:`cyclic isolate`, Python will not
+     automatically clear it if any member of the cyclic isolate has not yet
+     been automatically finalized (:c:member:`~PyTypeObject.tp_finalize`).
+   * Python will not destroy an object until after any automatic calls to its
+     :c:member:`!tp_clear` function have returned.  This ensures that the act
+     of breaking a reference cycle does not invalidate the ``self`` pointer
+     while :c:member:`!tp_clear` is still executing.
+   * Python will not automatically call :c:member:`!tp_clear` multiple times
+     concurrently.
+
+   Python currently only automatically clears objects as needed to break
```

Comment:
As a rule tof thumb, things that may break in future versions are CPython-specific.
```suggestion
   CPython currently only automatically clears objects as needed to break
```